ext/drmgr/drmgr.c

```diff
@@ -1194,6 +1194,33 @@ drmgr_is_last_instr(void *drcontext, instr_t *instr)
     return instr == pt->last_instr;
 }
 
+DR_EXPORT
+size_t
+drmgr_get_bb_size(instrlist_t *bb)
+{
+    instr_t *instr;
+    size_t size = 0;
+
+    for (instr = instrlist_first(bb); instr != NULL; instr = instr_get_next(instr))
+        size++;
+
+    return size;
+}
+
+DR_EXPORT
+size_t
+drmgr_get_bb_app_size(instrlist_t *bb)
```

Comment:
Hmm, users of drmgr who want this value should probably compute it in the analysis phase and remember it.  The drmgr usage model discourages walking the list in the insert phase.  The use case in the DR repo I can think of (aside from bbsize) is in drcachesim's instru_offline.cpp where it does just that: it counts in bb_analysis() and remembers the value (it can't use this function though b/c it has complexities in its count).

I guess what I'm getting at is, is providing this function going to tempt people to call it in the insert phase (esp since it was put in drmgr, rather than say drx, and the docs don't say which phases to use it in), or tempt them to call it many times thinking that drmgr is caching sthg for them.

Should it instead be drx_instrlist_size() and drx_instrlist_app_size() (or s/size/count/ which seems more natural at first to me but we have "bbsize" and STL container .size() so maybe best to stick w/ that) and not be endorsed by drmgr or specific to a bb or a phase?  Should the docs say that it is not O(1)?

Not a big deal, I don't feel strongly: just talking out loud.  If you prefer this let's go with this (maybe with the docs augmented to say "not meant to be called in the insert phase: cache it during the analysis phase instead" or sthg).

Reply:
I added that function b/c there are meta-instructions which are added during the app-2-app stage that one might want to take into account.

Reply:
hmm, I see your points. Let me think about it...

Reply:
So I do agree that drx is actually a better fit than drmgr. The only **minor** downside I see is that one will need to stay linking drx to use these function, which I do see as a bit of an overkill. Personally, I would have added these functions along with the other instrlist_* functions because they are all related and therefore more straight-forward for users. However, I also understand your point that we should keep the core small.

With regards to the issue of not counting the number of instructions of a bb during the insert stage, I see this as orthogonal, i.e., with or without these functions, the user should implicitly know that count is discouraged. I guess your point is that by including these functions in drmgr, then such counting is encouraged. A comment in the docs can alleviate this imo.

Reply:
@derekbruening Any idea on the above? I am not sure what is in agreement.

Reply:
With the comment this seems ok.

Reply:
Maybe get a 3rd opinion?  To avoid settling for what we seem to agree is not the best choice: in drmgr, which is all about mediating among events, which this seems unrelated to.

So we have my vote for drx.  Re: having to link drx: these are optional utility functions, which is exactly what drx is for.    We have a bunch of utility functions there.

And it sounds like your vote is to put into the core in instrlist.h.

Reply:
@AssadHashmi , @Carrotman42 : looking for opinions on where to place this feature

Reply:
I also don't have a strong opinion, but drx seems appropriate; like Derek mentioned, it's already a collection of helpers.

Reply:
Seems like drx it is then! I'll do the modifications. Thanks

Reply:
Done